### DISCUSSION TITLE
Zero-to-many relationship without foreign keys causes mapper to fail to initialize

### DISCUSSION THREAD

**Discussion by lunalucadou** [sqlalchemy/sqlalchemy#11251] [Usage Questions]
I have the following classes:

```python
class Permission(Base):
    __tablename__ = "permission"
    id: Mapped[int] = mapped_column(primary_key=True)
    app_id: Mapped[str]
    permission: Mapped[str]
```

```python
class Webhook(Base):
    __tablename__ = "webhook"
    id: Mapped[str] = mapped_column(primary_key=True)
    # ...
    app_id: Mapped[str]

    app_permissions: Mapped[list[Permission]] = relationship(
        Permission,
        primaryjoin="Webhook.app_id==Permission.app_id",
    )
```

```python
class Export(Base):
    __tablename__ = "export"
    id: Mapped[str] = mapped_column(primary_key=True)
    # ...
    app_id: Mapped[str]

    app_permissions: Mapped[list[Permission]] = relationship(
        Permission,
        primaryjoin="Export.app_id==Permission.app_id",
    )
```

`app_id` is not used as a foreign key because an app may have zero to many Permissions.

If I'm reading [this documentation](https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#specifying-alternate-join-conditions) correctly, I believe this should work. But when I try and run my code, the following errors occur:

```python
sqlalchemy.exc.InvalidRequestError: When initializing mapper Mapper[Webhook(webhook)], expression 'Webhook.app_id==Permission.app_id' failed to locate a name ("name 'Permission' is not defined"). If this is a class name, consider adding this relationship() to the <class 'webhook.Webhook'> class after both dependent classes have been defined.
sqlalchemy.exc.InvalidRequestError: When initializing mapper Mapper[Export(export)], expression 'Export.app_id==Permission.app_id' failed to locate a name ("name 'Permission' is not defined"). If this is a class name, consider adding this relationship() to the <class 'export.Export'> class after both dependent classes have been defined.
```

I have tried specifying `foreign_keys=[app_id]`, but the same error occurs. I have also verified that the classes are being imported properly.

However, there is a workaround, using lambdas:

```python
class Webhook(Base):
    __tablename__ = "webhook"
    id: Mapped[str] = mapped_column(primary_key=True)
    # ...
    app_id: Mapped[str]

    app_permissions: Mapped[list[Permission]] = relationship(
        Permission,
        primaryjoin=lambda: Webhook.app_id == foreign(Permission.app_id),
    )
```

```python
class Export(Base):
    __tablename__ = "export"
    id: Mapped[str] = mapped_column(primary_key=True)
    # ...
    app_id: Mapped[str]

    app_permissions: Mapped[list[Permission]] = relationship(
        Permission,
        primaryjoin=lambda: Export.app_id == foreign(Permission.app_id),
    )
```

I'm not sure why this works while my original code does not. Is this intended behavior?

**Comment by zzzeek** [comment]
That's not the error the code produces.  It produces this error:

    sqlalchemy.exc.ArgumentError: Could not locate any relevant foreign key columns for primary join condition 'webhook.app_id = permission.app_id' on relationship Webhook.app_permissions.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are annotated in the join condition with the foreign() annotation.

you can add the foreign annotation to the string directly:

    app_permissions: Mapped[list[Permission]] = relationship(
        Permission,
        primaryjoin="Webhook.app_id==foreign(Permission.app_id)",
    )

then you get the next problem which is:

     /home/classic/dev/sqlalchemy/test3.py:43: SAWarning: relationship 'Export.app_permissions' will copy column export.app_id to column permission.app_id, which conflicts with relationship(s): 'Webhook.app_permissions' (copies webhook.app_id to permission.app_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="app_permissions"' to the 'Export.app_permissions' relationship. (Background on this warning at: https://sqlalche.me/e/20/qzyx)


which is also correct.   How do you intend to load lists like Webhook.app_permissions while not also getting Export.app_permissions with the same app_id present?
